src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,1266 @@
+	 *     @type int|null    $match_offset Find the Nth tag matching all search criteria.
+	 *                                     1 for "first" tag, 3 for "third," etc.
+	 *                                     Defaults to first tag.
+	 *     @type string|null $class_name   Tag must contain this whole class name to match.
+	 *     @type string[]    $breadcrumbs  DOM sub-path at which element is found, e.g. `array( 'FIGURE', 'IMG' )`.
+	 * }
+	 * @return bool Whether a tag was matched.
+	 */
+	public function next_tag( $query = null ) {
+		if ( null === $query ) {
+			return $this->step();
+		}
+
+		if ( is_string( $query ) ) {
+			$query = array( 'breadcrumbs' => array( $query ) );
+		}
+
+		if ( ! is_array( $query ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Please pass a query array to this function.' ),
+				'6.4.0'
+			);
+			return false;
+		}
+
+		if ( ! ( array_key_exists( 'breadcrumbs', $query ) && is_array( $query['breadcrumbs'] ) ) ) {
+			return $this->step();
+		}
+
+		if ( isset( $query['tag_closers'] ) && 'visit' === $query['tag_closers'] ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Cannot visit tag closers in HTML Processor.' ),
+				'6.4.0'
+			);
+			return false;
+		}
+
+		$breadcrumbs  = $query['breadcrumbs'];
+		$match_offset = isset( $query['match_offset'] ) ? (int) $query['match_offset'] : 1;
+
+		$crumb  = end( $breadcrumbs );
+		$target = strtoupper( $crumb );
+		while ( $this->step() ) {
+			if ( $target !== $this->get_tag() ) {
+				continue;
+			}
+
+			// Look up the stack to see if the breadcrumbs match.
+			foreach ( $this->state->stack_of_open_elements->walk_up() as $node ) {
+				if ( strtoupper( $crumb ) !== $node->node_name ) {
+					break;
+				}
+
+				$crumb = prev( $breadcrumbs );
+				if ( false === $crumb && 0 === --$match_offset ) {
+					return true;
+				}
+			}
+
+			$crumb = end( $breadcrumbs );
+		}
+
+		return false;
+	}
+
+	/**
+	 * Steps through the HTML document and stop at the next tag, if any.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @throws Exception
+	 *
+	 * @see self::PROCESS_NEXT_NODE
+	 * @see self::REPROCESS_CURRENT_NODE
+	 *
+	 * @param string $node_to_process Whether to parse the next node or reprocess the current node.
+	 * @return bool Whether a tag was matched.
+	 */
+	public function step( $node_to_process = self::PROCESS_NEXT_NODE ) {
+		if ( self::PROCESS_NEXT_NODE === $node_to_process ) {
+			$top_node = $this->state->stack_of_open_elements->current_node();
+			if ( $top_node && self::is_void( $top_node->node_name ) ) {
+				$this->state->stack_of_open_elements->pop();
+			}
+
+			parent::next_tag( self::VISIT_EVERYTHING );
+		}
+
+		// Finish stepping when there are no more tokens in the document.
+		if ( null === $this->get_tag() ) {
+			return false;
+		}
+
+		$this->current_token = new WP_HTML_Token(
+			$this->bookmark_tag(),
+			$this->get_tag(),
+			$this->is_tag_closer(),
+			$this->release_internal_bookmark
+		);
```

Comment:
this is a tough one, and I'm not sure I entirely follow the association with `$lexical_updates` and `$classname_updates`

I've always had a version of the current token because so much of the parsing algorithm depends on properties of the current element/node/token. it has made sense to make it globally available within the processor.

of note, it only gets set in the one place that all scanning takes place (everything flows through `$this->step()`) and ultimately it's just a bookmark.

I'd like to better understand the ways this could get out of sync. maybe it does when seeking backwards? I will try and think up a test to confirm that.

Reply:
in f67e051531 I added a new test which I actually though _would_ break, but it seems to be fine. I don't have solid confidence that it's not still possible to break though.